File: interfaces/matlab/acado/packages/+acado/@Power/Power.m
%Power of expressions.
%
%  Usage:
%    >> Power(obj1, obj2);
%    >> obj1^obj2;
%
%  Parameters:
%    obj1 	    [Expression]
%    obj2       [Expression]
%
%  Licence:
%    This file is part of ACADO Toolkit  - (http://www.acadotoolkit.org/)
%
%    ACADO Toolkit -- A Toolkit for Automatic Control and Dynamic Optimization.
%    Copyright (C) 2008-2009 by Boris Houska and Hans Joachim Ferreau, K.U.Leuven.
%    Developed within the Optimization in Engineering Center (OPTEC) under
%    supervision of Moritz Diehl. All rights reserved.
%
%    ACADO Toolkit is free software; you can redistribute it and/or
%    modify it under the terms of the GNU Lesser General Public
%    License as published by the Free Software Foundation; either
%    version 3 of the License, or (at your option) any later version.
%
%    ACADO Toolkit is distributed in the hope that it will be useful,
%    but WITHOUT ANY WARRANTY; without even the implied warranty of
%    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
%    Lesser General Public License for more details.
%
%    You should have received a copy of the GNU Lesser General Public
%    License along with ACADO Toolkit; if not, write to the Free Software
%    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
%
%    Author: David Ariens, Rien Quirynen
%    Date: 2012
%
classdef Power < acado.BinaryOperator
    properties(SetAccess='private')

    end
    
    methods
        function obj = Power(obj1, obj2)
            if (isa(obj1, 'numeric'))
                obj1 = acado.DoubleConstant(obj1);
            elseif (isa(obj2, 'numeric'))
                obj2 = acado.DoubleConstant(obj2);
            end
            obj1 = obj1.getExpression;
            obj2 = obj2.getExpression;
            
            if obj1.zero
               obj.zero = 1;
            elseif obj2.zero
               obj.one = 1; 
            elseif obj1.one
               obj.one = 1;
            end
            
            obj.obj1 = obj1;
            obj.obj2 = obj2;
            obj.singleTerm = obj1.singleTerm && obj2.singleTerm;
        end
        
        function out = copy(obj)
            out = acado.Power(copy(obj.obj1), copy(obj.obj2));
        end
        
        function s = toString(obj)
            global ACADO_;
            
            if obj.obj1.zero
               obj.zero = 1;
            elseif obj.obj2.zero || obj.obj1.one
               obj.one = 1; 
            end
            
            if obj.zero
                s = '0';
            elseif obj.one
                s = '1';
            elseif strcmp(obj.obj2.toString, '1')
                s = sprintf('%s', obj.obj1.toString); 
            elseif isa(obj.obj2, 'acado.DoubleConstant') && obj.obj2.val == -1
                s = sprintf('1/%s', obj.obj1.toString); 
            elseif isa(obj.obj2, 'acado.DoubleConstant') && obj.obj2.val == 1/2
                s = sprintf('sqrt(%s)', obj.obj1.toString);
            elseif isa(obj.obj2, 'acado.DoubleConstant') && obj.obj2.val == -1/2
                s = sprintf('1/sqrt(%s)', obj.obj1.toString);
            elseif ~isempty(ACADO_) && ACADO_.generatingCode
                s = sprintf('pow(%s,%s)', obj.obj1.toString, obj.obj2.toString); 
            else
                if obj.obj1.singleTerm && obj.obj2.singleTerm
                    s = sprintf('%s^%s', obj.obj1.toString, obj.obj2.toString); 
                elseif obj.obj1.singleTerm
                    s = sprintf('%s^(%s)', obj.obj1.toString, obj.obj2.toString); 
                elseif obj.obj2.singleTerm
                    s = sprintf('(%s)^%s', obj.obj1.toString, obj.obj2.toString); 
                else
                    s = sprintf('(%s)^(%s)', obj.obj1.toString, obj.obj2.toString); 
                end
            end
        end
        
        function jac = jacobian(obj, var)
            if ~isvector(obj)
                error('A jacobian can only be computed of a vector function.');
            end
            for i = 1:length(obj)
                if obj(i).zero
                    jac(i,:) = zeros(1,length(var));
                elseif obj(i).one
                    jac(i,:) = zeros(1,length(var));
                elseif isa(obj(i).obj2, 'acado.DoubleConstant')
                    jac(i,:) = obj(i).obj2*acado.Power(obj(i).obj1,obj(i).obj2.val-1)*jacobian(obj(i).obj1, var);
                else
                    jac(i,:) = obj(i)*(jacobian(obj(i).obj1,var)*(obj(i).obj2/obj(i).obj1) + jacobian(obj(i).obj2,var)*acado.Logarithm(obj(i).obj1));
                end
            end
        end
        
        function out = simplifyLocally(obj)
            if isa(obj.obj1, 'acado.DoubleConstant') && isa(obj.obj2, 'acado.DoubleConstant')
                out = acado.DoubleConstant(obj.obj1.val^obj.obj2.val);
            else
                out = obj;
            end
        end
    end
    
end

File: src/symbolic_operator/power.cpp
/*
 *    This file is part of ACADO Toolkit.
 *
 *    ACADO Toolkit -- A Toolkit for Automatic Control and Dynamic Optimization.
 *    Copyright (C) 2008-2014 by Boris Houska, Hans Joachim Ferreau,
 *    Milan Vukov, Rien Quirynen, KU Leuven.
 *    Developed within the Optimization in Engineering Center (OPTEC)
 *    under supervision of Moritz Diehl. All rights reserved.
 *
 *    ACADO Toolkit is free software; you can redistribute it and/or
 *    modify it under the terms of the GNU Lesser General Public
 *    License as published by the Free Software Foundation; either
 *    version 3 of the License, or (at your option) any later version.
 *
 *    ACADO Toolkit is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *    Lesser General Public License for more details.
 *
 *    You should have received a copy of the GNU Lesser General Public
 *    License along with ACADO Toolkit; if not, write to the Free Software
 *    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 *
 */



/**
 *    \file src/symbolic_operator/power.cpp
 *    \author Boris Houska, Hans Joachim Ferreau
 *    \date 2008
 */


#include <acado/utils/acado_utils.hpp>
#include <acado/symbolic_operator/symbolic_operator.hpp>



BEGIN_NAMESPACE_ACADO




Power::Power( ):BinaryOperator( ){ }

Power::Power( const SharedOperator &_argument1, const SharedOperator &_argument2 )
      :BinaryOperator( _argument1, _argument2 ){ }

Power::Power( const Power &arg ):BinaryOperator( arg ){ }

Power::~Power(){ }

returnValue Power::evaluate( EvaluationBase *x ){

    x->power(*a1,*a2);
    return SUCCESSFUL_RETURN;
}


SharedOperator Power::substitute( SharedOperatorMap &sub ){

    return SharedOperator( new Power( a1->substitute(sub),
                                      a2->substitute(sub) ) );
}


std::ostream& Power::print( std::ostream &stream, StringMap &name ) const{

	if ( acadoIsEqual( a2->getValue(),0.5 ) == BT_TRUE )
	{
		stream << "(sqrt(";
		a1->print(stream,name);
		stream << "))"; 
		return  stream;
	}
	else
	{
		if ( acadoIsEqual( a2->getValue(),-0.5 ) == BT_TRUE )
		{
			stream << "(1.0/sqrt(";
			a1->print(stream,name);
			stream << "))"; 
			return  stream;
		}
		else
		{
			stream << "(pow(";
			a1->print(stream,name);
			stream << ",";
			a2->print(stream,name);
			stream << "))";
			return  stream;
		}
	}
}


returnValue Power::initDerivative() {

    if( d1 != 0 ) { return SUCCESSFUL_RETURN; }

    SharedOperator oneTmp  = SharedOperator( new DoubleConstant(1.0, NE_ONE));  // 1
    SharedOperator yminone = mySubtract( a2, oneTmp );                          // y-1

    SharedOperator xpowy       = convert2TreeProjection( myPower(a1,a2)     );  // x^y
    SharedOperator xpowyminone = convert2TreeProjection( myPower(a1,yminone));  // x^{y-1}
    SharedOperator logx        = convert2TreeProjection( myLogarithm(a1)    );  // log(x)

    d1 = convert2TreeProjection(myProd(xpowyminone,a2));                        // x^{y-1}*y
    d2 = convert2TreeProjection(myProd(logx,xpowy));                            // log(x)*x^y 

    SharedOperator twoTmp       = SharedOperator( new DoubleConstant(2.0,NE_NEITHER_ONE_NOR_ZERO)); // 2
    SharedOperator subTmp2      = mySubtract( a2, twoTmp );                                         // y-2
    SharedOperator prodTmp      = myProd( a2, yminone );                                            // y*(y-1)
    SharedOperator prodTmp2     = myProd( a2, logx );                                               // y*log(x)
    SharedOperator oneplusylogx = myAdd( oneTmp, prodTmp2 );                                        // 1+y*log(x)
    SharedOperator xpowymintwo  = myPower( a1, subTmp2);                                            // x^(y-2)

    d11 = convert2TreeProjection( myProd( prodTmp, xpowymintwo      ) );                            // y*(y-1)*x^(y-2)
    d12 = convert2TreeProjection( myProd( oneplusylogx, xpowyminone ) );                            // (1+y*log(x))*x^(y-1)
    d22 = convert2TreeProjection( myProd(logx,d2) );                                                // log(x)^2*x^y
    
    a1->initDerivative();
    return a2->initDerivative();
}


CLOSE_NAMESPACE_ACADO

// end of file.
